### PR TITLE
[SPARK-3000] Allow MemoryStore blocks to be dropped to disk in parallel

### PR DESCRIPTION
**Pull request by JoshRosen**
This patch refactors the BlockManager and MemoryManager locking/synchronization logic in order to allow MemoryStore blocks to be dropped to the disk in parallel, significantly reducing `MemoryManager` lock contention when acquiring storage memory.
## Key change

Prior to this patch, `MemoryStore.evictBlocksToFreeSpace()` synchronized on `memoryManager` for the entire block-dropping process. This patch reduces the scope of the synchronization to only cover the phase in which we are selecting the set of blocks to be evicted. Once a task has committed to dropping a set of blocks, it now releases the `memoryManager` lock and performs the block eviction without holding any locks. This allows multiple tasks to drop blocks to disk in parallel and avoids blocking non-spilling tasks via contention on the `memoryManager` lock.

One tricky concern is preventing tasks from stealing the freed storage memory which results from eviction. To address this, this patch modifies the eviction path so that evicted blocks' memory is not immediately freed with the memory manager. Once all of a tasks's block evictions have completed, tasks now compare the total amount of memory freed from eviction (which could exceed their request size) to the amount of memory needed and either acquire more memory or free excess storage memory.

Another concern is making sure that a single block isn't double-evicted by multiple tasks. Double-eviction is prevented by the fact that tasks obtain exclusive write locks on all of the blocks that they will evict and the fact that the process of selecting which blocks to evict is still synchronized on the `memoryManager`.

Note that a single task will still spill blocks sequentially, so parallelism only results from multiple tasks triggering separate block evictions. If we want to enable a single task's evictions to proceed in parallel then we can do that in a separate followup patch (it should be a fairly straightforward change).
## Motivating performance results

This patch is motivated by a simple `spark-shell` benchmark that I wrote to stress the BlockManager's block eviction path:

``` scala
import org.apache.spark.storage.StorageLevel
val startTime = System.currentTimeMillis
(1 to 100).map { _ => 
  sc.parallelize(1 to 1000, 100)
    .map(x => new Array[Byte](1024 * 1024 / 10))
    .persist(StorageLevel.MEMORY_AND_DISK)
}.map { x => x.count(); x }
val endTime = System.currentTimeMillis
println((endTime - startTime) / 1000.0)
```

---

**Prior to this patch**, I observed extreme contention for the `UnifiedMemoryManager`'s monitor:

image

image

---

**After this patch**, the contention on the `UnifiedMemoryManager` lock is significantly reduced and the lock contention bottlenecks shifted to other sources:

image

image

### CODE REVIEW

**Comment by rxin** [comment]
cc @sameeragarwal too

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

File: core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala [hunk 1]
```diff
@@ -316,17 +324,42 @@ private[spark] class MemoryStore(
     *
     * @param blockId the ID of the block we are freeing space for, if any
     * @param space the size of this block
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala]: This variable name / comment is outdated; `space` is the amount of memory that we'd like to free via eviction, which is not necessarily equal to the whole size of the block.

File: core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala [hunk 2]
```diff
@@ -56,11 +56,13 @@ private[memory] class StorageMemoryPool(lock: Object) extends MemoryPool(lock) w
...
+   *
    * @return whether all N bytes were successfully granted.
    */
-  def acquireMemory(blockId: BlockId, numBytes: Long): Boolean = lock.synchronized {
-    val numBytesToFree = math.max(0, numBytes - memoryFree)
+  def acquireMemory(blockId: BlockId, numBytes: Long): Boolean = {
+    val numBytesToFree = lock.synchronized {
+      math.max(0, numBytes - memoryFree)
+    }
     acquireMemory(blockId, numBytes, numBytesToFree)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala]: As discussed, the fact that we release the lock here could lead to two race conditions:

(1) If another thread comes in and uses up some free memory before this line, then `numBytesToFree` will be too low, causing the memory acquisition to fail.

(2) If another thread comes in and frees up a lot of memory before this line, then `numBytesToFree` will be too high, causing more blocks to be evicted than necessary.

We can remedy (1) by acquiring as much free space is needed and available before releasing the lock and then try to acquire the rest later. Fixing (2) may be a little more difficult.

File: core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala [hunk 3]
```diff
@@ -206,11 +206,12 @@ private[spark] class MemoryStore(
       }
       val size = entry.size
       def transferUnrollToStorage(amount: Long): Unit = {
```
Comment (davies) [inline core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala]: Could you change this function name to reflect the change here?

**Comment by davies** [comment]
It took me long time to understand these changes, LGTM. Fine-gain locking is usually hard to prove, it's better to have another people also review this.

File: core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala (hunk 3 above)
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala]: I'm actually going to just revert this change since it's fine to synchronize this transfer since the transfer is incapable of spilling.

File: core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala (hunk 2 above)
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala]: I just pushed a patch which aims to address (1). I don't think that (2) is a concern because we're only missing out on an opportunity which did not exist prior to this patch's changes. In the old code, other tasks' calls to free storage memory would be blocked due to contention on the MemoryManager lock, so this transaction wouldn't observe the effect of free()s which took place after it started.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

File: core/src/main/scala/org/apache/spark/memory/StaticMemoryManager.scala [hunk 4]
```diff
@@ -62,17 +64,34 @@ private[spark] class StaticMemoryManager(
...
-    storageMemoryPool.acquireMemory(blockId, numBytes, numBytesToFree)
+    val (freeMemoryAcquired, numBytesToFree) = synchronized {
+      val currentUnrollMemory = storageMemoryPool.memoryStore.currentUnrollMemory
+      val freeMemory = storageMemoryPool.memoryFree
+      val freeMemoryAcquired = math.min(numBytes, freeMemory)
+      val res = storageMemoryPool.acquireMemory(blockId, freeMemoryAcquired, numBytesToFree = 0)
+      assert(res, "failed to acquire free storage memory")
+      val maxNumBytesToFree = math.max(0, maxUnrollMemory - currentUnrollMemory - freeMemory)
+      // Keep it within the range 0 <= X <= maxNumBytesToFree
+      val numBytesToFree = math.min(numBytes - freeMemoryAcquired, maxNumBytesToFree)
```
Comment (sameeragarwal) [inline core/src/main/scala/org/apache/spark/memory/StaticMemoryManager.scala]: Out of curiosity, what happens if `numBytesToFree < numBytes - freeMemoryAcquired`? Do we then use the disk to unroll the block?

File: core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala [hunk 5]
```diff
@@ -75,21 +95,31 @@ private[memory] class StorageMemoryPool(lock: Object) extends MemoryPool(lock) w
...
     }
-    // NOTE: If the memory store evicts blocks, then those evictions will synchronously call
-    // back into this StorageMemoryPool in order to free memory. Therefore, these variables
-    // should have been updated.
-    val enoughMemory = numBytesToAcquire <= memoryFree
-    if (enoughMemory) {
-      _memoryUsed += numBytesToAcquire
+    if (memoryFreedByEviction > numBytesToAcquire) {
+      releaseMemory(memoryFreedByEviction - numBytesToAcquire)
+    }
```
Comment (sameeragarwal) [inline core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala]: maybe put the remaining code below in an `else` block?

File: core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala [hunk 6]
```diff
@@ -131,26 +131,28 @@ private[spark] class UnifiedMemoryManager private[memory] (
...
+        logInfo(s"Will not store $blockId as the required space ($numBytes bytes) exceeds our " +
+          s"memory limit ($maxStorageMemory bytes)")
+        return false
+      }
+      if (numBytes > storageMemoryPool.memoryFree) {
+        // There is not enough free memory in the storage pool, so try to borrow free memory from
+        // the execution pool.
+        val memoryBorrowedFromExecution = Math.min(onHeapExecutionMemoryPool.memoryFree, numBytes)
+        onHeapExecutionMemoryPool.decrementPoolSize(memoryBorrowedFromExecution)
+        storageMemoryPool.incrementPoolSize(memoryBorrowedFromExecution)
```
Comment (sameeragarwal) [inline core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala]: out of curiosity, why do we synchronize on the whole block instead of just these 2 lines above?

**Comment by sameeragarwal** [comment]
Not really a review as I was just trying to understand the MemoryManager, but overall looks great. Do we have unit tests that try to acquire storage/unroll memory in parallel?

File: core/src/main/scala/org/apache/spark/memory/StaticMemoryManager.scala (hunk 4 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/memory/StaticMemoryManager.scala]: can we be sure that `numBytesToFree` won't be negative? Did you remove the `math.max(0, ...)` on purpose?

File: core/src/main/scala/org/apache/spark/memory/StaticMemoryManager.scala [hunk 7]
```diff
@@ -62,17 +64,34 @@ private[spark] class StaticMemoryManager(
...
+    try {
+      if (storageMemoryPool.acquireMemory(blockId, numBytes - freeMemoryAcquired, numBytesToFree)) {
+        true
+      } else {
+        storageMemoryPool.releaseMemory(freeMemoryAcquired)
+        false
+      }
+    } catch {
+      case NonFatal(e) =>
+        storageMemoryPool.releaseMemory(freeMemoryAcquired)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/memory/StaticMemoryManager.scala]: it's kind of weird that it's possible to call `releaseMemory` twice here. If the first one fails at the end (after actually releasing memory) it will come here and possibly release the memory again (and then failing again). A better way to write this would be something like:
```
// Acquire the rest of the memory outside a synchronized block.
// This call may evict existing blocks if necessary.
val acquired =
  try {
    storageMemoryPool.acquire(...)
  } catch {
    case NonFatal(e) =>
      storageMemoryPool.releaseMemory(...)
      throw e
  }
if (!acquired) {
  storageMemoryPool.releaseMemory(...)
}
acquired
```

File: core/src/main/scala/org/apache/spark/memory/StaticMemoryManager.scala [hunk 8]
```diff
@@ -62,17 +64,34 @@ private[spark] class StaticMemoryManager(
...
+      val freeMemoryAcquired = math.min(numBytes, freeMemory)
+      val res = storageMemoryPool.acquireMemory(blockId, freeMemoryAcquired, numBytesToFree = 0)
+      assert(res, "failed to acquire free storage memory")
+      val maxNumBytesToFree = math.max(0, maxUnrollMemory - currentUnrollMemory - freeMemory)
+      // Keep it within the range 0 <= X <= maxNumBytesToFree
+      val numBytesToFree = math.min(numBytes - freeMemoryAcquired, maxNumBytesToFree)
+      (freeMemoryAcquired, numBytesToFree)
+    }
+    try {
+      if (storageMemoryPool.acquireMemory(blockId, numBytes - freeMemoryAcquired, numBytesToFree)) {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/memory/StaticMemoryManager.scala]: actually, don't you want to check whether `numBytes - freeMemoryAcquired` is > 0? If it's 0 then we should probably just return true here without calling anything.

File: core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala [hunk 9]
```diff
@@ -56,12 +58,30 @@ private[memory] class StorageMemoryPool(lock: Object) extends MemoryPool(lock) w
...
    * Acquire N bytes of memory to cache the given block, evicting existing ones if necessary.
- *
+   *
    * @return whether all N bytes were successfully granted.
    */
-  def acquireMemory(blockId: BlockId, numBytes: Long): Boolean = lock.synchronized {
-    val numBytesToFree = math.max(0, numBytes - memoryFree)
-    acquireMemory(blockId, numBytes, numBytesToFree)
+  def acquireMemory(blockId: BlockId, numBytes: Long): Boolean = {
+    // First, attempt to acquire as much memory as we can without performing any eviction
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala]: I would expand on this: "Otherwise, another thread may jump in and use up some free memory as soon as we release the lock, in which case we may evict too few blocks to store this block later."

Right now it's not clear why we need to do this in 2 steps.

File: core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala [hunk 10]
```diff
@@ -56,12 +58,30 @@ private[memory] class StorageMemoryPool(lock: Object) extends MemoryPool(lock) w
...
+    val bytesAcquiredWithoutEviction = lock.synchronized {
+      val bytesAcquired = math.min(numBytes, memoryFree)
+      _memoryUsed += bytesAcquired
+      bytesAcquired
+    }
+    // Acquire the remainder of the memory by performing eviction
+    try {
+      val bytesToAcquireViaEviction = numBytes - bytesAcquiredWithoutEviction
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala]: can you add some asserts here to make sure this is non-negative? Also I would do all of this only if this is > 0.

File: core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala [hunk 11]
```diff
@@ -75,21 +95,31 @@ private[memory] class StorageMemoryPool(lock: Object) extends MemoryPool(lock) w
   def acquireMemory(
       blockId: BlockId,
       numBytesToAcquire: Long,
-      numBytesToFree: Long): Boolean = lock.synchronized {
+      numBytesToFree: Long): Boolean = {
     assert(numBytesToAcquire >= 0)
     assert(numBytesToFree >= 0)
-    assert(memoryUsed <= poolSize)
-    if (numBytesToFree > 0) {
+    assert(lock.synchronized(memoryUsed <= poolSize))
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala]: nit can you do
```
lock.synchronized { assert(memoryUsed <= poolSize) }

File: core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala [hunk 12]
```diff
@@ -56,12 +58,30 @@ private[memory] class StorageMemoryPool(lock: Object) extends MemoryPool(lock) w
...
-    val numBytesToFree = math.max(0, numBytes - memoryFree)
-    acquireMemory(blockId, numBytes, numBytesToFree)
+  def acquireMemory(blockId: BlockId, numBytes: Long): Boolean = {
+    // First, attempt to acquire as much memory as we can without performing any eviction
+    val bytesAcquiredWithoutEviction = lock.synchronized {
+      val bytesAcquired = math.min(numBytes, memoryFree)
+      _memoryUsed += bytesAcquired
+      bytesAcquired
+    }
+    // Acquire the remainder of the memory by performing eviction
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala]: can you add something like
```
// Here we want to evict as much space as we are requesting because we have already
// acquired all available free space for this block.
```

File: core/src/main/scala/org/apache/spark/memory/StaticMemoryManager.scala [hunk 13]
```diff
@@ -17,6 +17,8 @@
 
 package org.apache.spark.memory
 
+import scala.util.control.NonFatal
+
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/memory/StaticMemoryManager.scala]: by the way I'm not sure whether it's worth it to change anything here. This is for legacy mode and will be removed in the future anyway, and for those people who *are* using it they're doing so because the new workload doesn't work for them, so it's best to just keep things the same here as much as possible. I don't feel super strongly about reverting the changes here but I don't think it's totally necessary.

File: core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala [hunk 14]
```diff
@@ -56,12 +58,30 @@ private[memory] class StorageMemoryPool(lock: Object) extends MemoryPool(lock) w
...
+      val bytesToAcquireViaEviction = numBytes - bytesAcquiredWithoutEviction
+      if (acquireMemory(blockId, bytesToAcquireViaEviction, bytesToAcquireViaEviction)) {
+        true
+      } else {
+        releaseMemory(bytesAcquiredWithoutEviction)
+        false
+      }
+    } catch {
+      case NonFatal(e) =>
+        releaseMemory(bytesAcquiredWithoutEviction)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala]: same double release comment here

**Comment by JoshRosen** [comment]
FYI I'm working on a complete rewrite of this patch from first principles.

File: core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala (hunk 5 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala]: +1, also I'll just wrap both things in synchronized so we don't have to acquire the lock twice. Fewer interleavings to think about.

File: core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala (hunk 11 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala]: actually IDK if it's worth asserting this here if we have to acquire the lock for it. I'm OK with just removing it... (or moving it to release or something)

File: core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala [hunk 15]
```diff
@@ -110,17 +140,21 @@ private[memory] class StorageMemoryPool(lock: Object) extends MemoryPool(lock) w
    * Try to shrink the size of this storage memory pool by `spaceToFree` bytes. Return the number
    * of bytes removed from the pool's capacity.
    */
-  def shrinkPoolToFreeSpace(spaceToFree: Long): Long = lock.synchronized {
-    // First, shrink the pool by reclaiming free memory:
-    val spaceFreedByReleasingUnusedMemory = math.min(spaceToFree, memoryFree)
-    decrementPoolSize(spaceFreedByReleasingUnusedMemory)
-    val remainingSpaceToFree = spaceToFree - spaceFreedByReleasingUnusedMemory
+  def shrinkPoolToFreeSpace(spaceToFree: Long): Long = {
+    var spaceFreedByReleasingUnusedMemory = 0L
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala]: if you do the `val (x, y) = lock.synchronized { ... }` thing here then you don't need a var

File: core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala [hunk 16]
```diff
@@ -315,6 +315,13 @@ private[spark] class MemoryStore(
     }
   }
 
+  private[storage] def removeBlockAndKeepStorageMemory(blockId: BlockId): Long = {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala]: this is too much... just add a boolean flag to the existing `remove`

File: core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala [hunk 17]
```diff
@@ -410,9 +414,23 @@ private[spark] class MemoryStore(
         selectedBlocks.foreach { id =>
           blockInfoManager.unlock(id)
         }
-        0L
+        selectedBlocks.clear()
+        freedMemory = 0
       }
     }
+
+    // At this point, `selectedBlocks` is the (possibly-empty) set of blocks that we have chosen
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala]: possibly empty

File: core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala (hunk 6 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala]: everything in this block is cheap anyway. The only expensive thing we want to keep out of the synchronized block here is evicting blocks from memory store, which is what `storageMemoryPool.acquireMemory` does.

File: core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala [hunk 18]
```diff
@@ -338,24 +345,51 @@ private[spark] class MemoryStore(
     * RDDs that don't fit into memory that we want to avoid).
     *
     * @param blockId the ID of the block we are freeing space for, if any
-    * @param space the size of this block
-    * @return the amount of memory (in bytes) freed by eviction
+    * @param requestedSpace the amount of memory which should be freed via block eviction
+    * @return the total size (in bytes) of the blocks evicted from the MemoryStore. This memory is
+    *         not released with the memory manager, so it's the caller's responsibility to perform
+    *         the appropriate bookkeeping calls.
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala]: I don't understand why this should be the case. I get that we don't want to acquire the lock for each block we drop, but can't we just release all the memory at the end of this method? We still only acquire 1 lock and now the caller doesn't have to worry about forgetting to release memory.

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala]: Is it because we want to transfer memory atomically to accommodate the new storage memory requests? Then we should say so in the comment.

**Comment by andrewor14** [comment]
I left a few suggestions on how to rewrite things in a way that's more readable. I'm not sure if we should change `StaticMemoryManager` since new users will probably not use that, and old users already suffer from the lock contention problem. Making changes there might introduce regressions and increase the surface area of things we need to maintain in Spark 2.0.

File: core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala (hunk 11 above)
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala]: This lock acquisition should be cheap, so I think it's fine to have this assert.

**Comment by JoshRosen** [comment]
I'm going to close this for now. Per some offline discussion, I think there's an easier way to implement all of this by significantly refactoring / simplifying MemoryManager.
